Add commands and queries to burrow integration

Ayush-Jalan opened the pull request:
<!-- You will not see HTML commented line in Pull Request body -->
<!-- Optional sections may be omitted. Just remove them or write None -->

<!-- ### Requirements -->
<!-- * Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion. -->
<!-- * All new code must have code coverage above 70% (https://docs.codecov.io/docs/about-code-coverage). -->
<!-- * CircleCI builds must be passed. -->
<!-- * Critical and blocker issues reported by Sorabot must be fixed. -->
<!-- * Branch must be rebased onto base branch (https://soramitsu.atlassian.net/wiki/spaces/IS/pages/11173889/Rebase+and+merge+guide). -->


### Description of the Change
Adds iroha commands and queries to burrow integration
Examples for the integration have also been added

<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. -->
<!-- Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->

### Benefits
This will extend HL Burrow integration into HL Iroha , so more commands can be used
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
All the commands and queries haven't been integrated like GrantPermission and GetAccountTransactions, etc.
<!-- What are the possible side-effects or negative impacts of the code change? -->
<!-- If no drawbacks, explicitly mention this (write None) -->

### Usage Examples or Tests *[optional]*
Examples have been added under /example/burrow_integration
<!-- Point reviewers to the test, code example or documentation which shows usage example of this feature -->



<!-- Explain what other alternates were considered and why the proposed version was selected -->

<!--
NOTE: User may want skip pull request and push workflows with [skip ci]
https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/
Phrases: [skip ci], [ci skip], [no ci], [skip actions], or [actions skip]
-->

baziorek commented on example/burrow_integration/account.py:
```diff
@@ -0,0 +1,183 @@
+    return str(number_hex).zfill(width)
+
+
+def left_padded_address_of_param(
```
The code looks quite good now, but the problem is that those functions are copy-paste in all files. I suggest to extract them into separate file, e.g. `integration_helpers.py` and import them from all files. Thanks to that all samples would be shorter

baziorek commented on example/burrow_integration/asset.py:
```diff
@@ -0,0 +1,258 @@
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # account id
+    params = params + argument_encoding("cc#burrow")  # asset id
```
I suggest to add global constant e.g.:
`ASSET_ID` instead of the same text in multiple places

baziorek commented on example/burrow_integration/detail.py:
```diff
@@ -0,0 +1,156 @@
+    params = params + argument_encoding("university")  # domain id
+    params = params + argument_encoding("MIT")  #  key
+    tx = iroha.transaction(
+        [iroha.command("CallEngine", caller="admin@test", callee=address, input=params)]
```
```suggestion
        [iroha.command("CallEngine", caller=ADMIN_ACCOUNT_ID, callee=address, input=params)]
```

baziorek commented on example/burrow_integration/detail.sol:
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: GPL-3.0
```
I suggest to change licence into the same as in examples: https://github.com/hyperledger/iroha-python/blob/41a724b3f3f21921ad455d39fb959f45c98fe986/examples/batch-example.py#L4
```
Apache-2.0
```

Ayush-Jalan replied:
What difference will it make?

baziorek replied:
> What difference will it make?

I'm not specialist, but as I know Apache is better than GPL, because when You use GPL code it makes Your code to be also GPL. Apache is not that sctict

baziorek commented on example/burrow_integration/integration_example.py:
```diff
@@ -0,0 +1,217 @@
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # account id
+    params = params + argument_encoding("coin#domain")  # asset id
```
I suggest to have global constant with domain

Ayush-Jalan replied:
It is not used so often, and mostly used in different forms.

baziorek commented on example/burrow_integration/iroha.sol:
```diff
@@ -0,0 +1,67 @@
+        result = ret;
+    }
+
+        //Transfers asset from one iroha account to another
+    function transferAsset(string memory src, string memory dst, string memory asset, string memory amount) public returns (bytes memory result) {
```
```suggestion
    //Transfers asset from one iroha account to another
    function transferAsset(string memory src, string memory dst, string memory asset, string memory amount) public returns (bytes memory result) {
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    bytecode = "608060405234801561001057600080fd5b5073a6abc17819738299b3b2c1ce46d55c74f04e290c6000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550610b2a806100746000396000f3fe608060405234801561001057600080fd5b506004361061004c5760003560e01c806368649379146100515780637fcd103614610081578063866bd458146100b1578063d4e804ab146100e15 ...
+    """Bytecode was generated using remix editor  https://remix.ethereum.org/ from file signatory.sol. """
+    tx = iroha.transaction(
+        [iroha.command("CallEngine", caller="admin@test", input=bytecode)]
```
```suggestion
        [iroha.command("CallEngine", caller=ADMIN_ACCOUNT_ID, input=bytecode)]
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    no_of_param = 2
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding("admin@test")  # account id
```
```suggestion
    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # account id
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    params = params + argument_encoding("admin@test")  # account id
+    params = params + argument_encoding(test_public_key)  # public key
+    tx = iroha.transaction(
+        [iroha.command("CallEngine", caller="admin@test", callee=address, input=params)]
```
```suggestion
        [iroha.command("CallEngine", caller=ADMIN_ACCOUNT_ID, callee=address, input=params)]
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    no_of_param = 2
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding("admin@test")  # account id
```
```suggestion
    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # account id
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    params = params + argument_encoding("admin@test")  # account id
+    params = params + argument_encoding(test_public_key)  # public key
+    tx = iroha.transaction(
+        [iroha.command("CallEngine", caller="admin@test", callee=address, input=params)]
```
```suggestion
        [iroha.command("CallEngine", caller=ADMIN_ACCOUNT_ID, callee=address, input=params)]
```

baziorek commented on example/burrow_integration/signatory.py:
```diff
@@ -0,0 +1,175 @@
+    no_of_param = 1
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding("admin@test")  # account id
+    tx = iroha.transaction(
+        [iroha.command("CallEngine", caller="admin@test", callee=address, input=params)]
```
```suggestion
    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # account id
    tx = iroha.transaction(
        [iroha.command("CallEngine", caller=ADMIN_ACCOUNT_ID, callee=address, input=params)]
```

baziorek commented on example/burrow_integration/detail.py:
```diff
@@ -0,0 +1,156 @@
+    no_of_param = 3
+    for x in range(no_of_param):
+        params = params + left_padded_address_of_param(x, no_of_param)
+    params = params + argument_encoding("admin@test")  # source account id
```
```suggestion
    params = params + argument_encoding(ADMIN_ACCOUNT_ID)  # source account id
```

baziorek reviewed (changes requested):
The PR is much better than previous one, You are learning fast. In current PR there are many changes, but vast majority of them can be corrected fast by applying suggestions (of course if You agree with them).

About my comments:
1. Extracting common function - it is for many examples.
2. I suggest to have the same name of contracts as in documentation: https://iroha.readthedocs.io/en/main/develop/api.html of course if possible
3. When You change api (adding description) please also change documentation: https://iroha.readthedocs.io/en/main/integrations/burrow_example.html sending in documentation doesn't have `description` argument.
4. Changes in documentation would be also necessarily, we don't have much time, so in the PR.
5. One more integration sample would be good: smart contract which is making multiple things at once, e.g.:
    1. createAccount
    2. setAccountDetails
    3. addAssetQuantity or transfering assets to the account (make initial balance)    4.

Ayush-Jalan commented on goSrc/src/vmCaller/evm/native_contract.go:
```diff
@@ -36,6 +38,208 @@ var (
+				* @return 'true' if successful, 'false' otherwise
+				`,
+			PermFlag: permission.Call,
+			F:        setDetail,
```
should I change the name in contract as well?

Ayush-Jalan commented on goSrc/src/vmCaller/iroha/commands.go:
```diff
@@ -82,6 +369,223 @@ func GetAccountAssets(accountID string) ([]*pb.AccountAsset, error) {
+func GetSignatories(accountID string) ([]string, error) {
+	query := &pb.Query{Payload: &pb.Query_Payload{
+		Meta: &pb.QueryPayloadMeta{
+			CreatedTime:      uint64(time.Now().UnixNano() / int64(time.Millisecond)),
```
This is how it was done in the previous integration.
And it will reduce only one line of code, as again QueryPayloadMeta is part of Query_Payload, so it will have to be in that form.

baziorek commented on example/burrow_integration/integration_helpers.py:
```diff
@@ -0,0 +1,89 @@
+def left_padded_address_of_param(
+    param_index: int, number_of_params: int, width: int = 64
+):
+    """Specifies the position of each argument according to Contract AbI specifications."""
```
```suggestion
    """Specifies the position of each argument according to Contract ABI specifications."""
```

baziorek reviewed (changes requested):
The PR is enough good to convert its from Draft into normal PR after correcting my not fixed comments.

Then please do few things:
1. When You change api (adding description) please also change documentation: https://iroha.readthedocs.io/en/main/integrations/burrow_example.html sending in documentation doesn't have description argument.
1. Changes in documentation would be also necessarily, we don't have much time, so in the PR.
1. One more integration sample would be good: smart contract which is making multiple things at once, e.g.:
    1. createAccount
    1. setAccountDetails
    1. addAssetQuantity or transfering assets to the account (to make initial balance for account)

baziorek commented on example/burrow_integration/setupAccount.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    function createAndSetupAccount(string memory adminAccountId, string memory userAccountName, string memory userAccountId, string memory assetId, string memory domain, string memory description, string memory amount, string memory key, string memory value, string memory publicKey) public returns (bytes memory result) {
+        result = createAccount(userAccountName, domain, publicKey);
+        result = transferAsset(adminAccountId, userAccountId, assetId, description, amount);
+        result = setAccountDetail(userAccountId, key, value);
```
I suggest to add checking: if result is error then don't call another function.

LiraLemur replied:
@baziorek @Ayush-Jalan This PR is rather big so I would suggest creating a new one for documentation and just referencing this one in the description. Please note that we usually start PR names for documentation with something like "Docs:...."
If you need any help with how documentation works or anything, please let me know in telegram group :)

Ayush-Jalan replied:
That is being checked for each function inside it, so I thought not to add it.

Ayush-Jalan replied:
@LiraLemur Okay, I will create a new PR for documentation.:)
If I am not wrong these are the steps to be followed for documentation ... https://github.com/hyperledger/iroha/tree/main/docs

baziorek replied:
>  If I am not wrong these are the steps to be followed for documentation ... https://github.com/hyperledger/iroha/tree/main/docs

Yes, You should:

1. checkout iroha
2. inside the main iroha directory: `cd docs/source`
3. `pip install -r requirements.txt`
4. `make html`
5. "After that, the documentation is generated in html format in _build folder"

LiraLemur replied:
@ayush-jalan you need to change the source files in docs/source - those are mostly .rst files. Please contact me in telegram if you have questions about that format of general questions on readthedocs documentation.

baziorek commented on example/burrow_integration/setupAccount.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    function setsAsset(string memory assetName, string memory domainName, string memory precision, string memory assetId, string memory amount) public returns (bytes memory result) {
+        result = createAsset(assetName, domainName, precision);
+        result = addAsset(assetId, amount);
```
as above

baziorek reviewed (approved):
Tho code looks good for me.

Now it is time for Iroha Core Team to review changes.

In meantime You can work on documentation - it can be in another PR, or this one.

CC: @LiraLemur

iceseer commented on goSrc/src/vmCaller/iroha/commands.go:
```diff
@@ -97,6 +97,248 @@ func AddAssetQuantity(asset string, amount string) error {
+}
+
+func CreateAsset(name string, domain string, precision string) error {
+	precision_uint, err := strconv.ParseUint(precision, 10, 32)
+	command := &pb.Command{Command: &pb.Command_CreateAsset{
+		CreateAsset: &pb.CreateAsset{
+			AssetName: name,
+			DomainId:  domain,
+			Precision: uint32(precision_uint),
+		}}}
+	commandResult, err := makeProtobufCmdAndExecute(IrohaCommandExecutor, command)
+	if err != nil {
+		return err
+	}
+	if commandResult.error_code != 0 {
+		error_extra := ""
+		error_extra_ptr := commandResult.error_extra.toStringAndRelease()
+		if error_extra_ptr != nil {
+			error_extra = ": " + *error_extra_ptr
+		}
+		return fmt.Errorf("Error executing CreateAsset command: %s", error_extra)
+	}
+
+	return nil
```
All command differs the only command structure and command name in log output. It seems can be refactored to a single implementation of this logic in some function. And calls of the  function elsewhere.

Also queries can be refactored the same way.

Ayush-Jalan replied:
@iceseer  Are you suggesting to have an error handling part(i.e. line 285 to 295 )in  a function?

baziorek replied:
> @iceseer Are you suggesting to have an error handling part(i.e. line 285 to 295 )in a function?

@Ayush-Jalan I think he suggested to extract function for all commands, because they are quite similar:
![image](https://user-images.githubusercontent.com/15332594/130181178-a84dcdc2-854d-4274-94cd-91a9d85b8ab6.png)
And call the function from all integrated commands.

Ayush-Jalan replied:
I have refactored the commands. Please, have a look 
@iceseer  @baziorek

Ayush-Jalan replied:
Refactoring for queries was not possible since they differ largely for each Query

Ayush-Jalan commented on goSrc/src/vmCaller/evm/native_contract.go:
```diff
@@ -154,19 +337,414 @@ type addAssetRets struct {
 		return addAssetRets{Result: false}, err
 	}
 
-	ctx.Logger.Trace.Log("function", "addAsset",
```
I have made the changes in the name

baziorek reviewed (approved):
It looks much better. I see that all suggestions from @iceseer was solved or explained why not.

kamilsa commented:
@Ayush-Jalan could you please fix integration_engine_call_test?

Ayush-Jalan commented:
> @Ayush-Jalan could you please fix integration_engine_call_test?

@kamilsa I haven't made any change to the integration_engine _call_test or related files. The checks were successful before the last commit.
I will try to look into it regarding what is the problem.

Ayush-Jalan commented:
@kamilsa  can it be failing because it says that this branch is out of date with the base branch? @baziorek

kamilsa commented:
> @kamilsa can it be failing because it says that this branch is out of date with the base branch? @baziorek

@Ayush-Jalan no. We will try to help with what is the issue

Ayush-Jalan commented:
@kamilsa I guess the reason for error is that I have changed the transferAsset commands in the integration and added one more argument, so the TransferAsset test is failing inside the call_engine_test .
